Native histograms: reconcile mismatched NHCB bounds in Add/Sub

linasm opened the pull request:
#### Which issue(s) does the PR fix:
Fixes #17255

#### Does this PR introduce a user-facing change?
```release-notes
`Add` and `Sub` methods called on NHCB histograms with different bucket bounds will no longer result in an error
(which was resulting in query warning). Instead, observation counts in unmatched buckets on either side will be 
rolled up into the next matching bucket (which could be the `+Inf` serving as the last, catch all bucket).
```

linasm commented on util/annotations/annotations.go:
```diff
@@ -134,16 +134,15 @@ var (
-	NativeHistogramNotCounterWarning           = fmt.Errorf("%w: this native histogram metric is not a counter:", PromQLWarning)
-	NativeHistogramNotGaugeWarning             = fmt.Errorf("%w: this native histogram metric is not a gauge:", PromQLWarning)
-	MixedExponentialCustomHistogramsWarning    = fmt.Errorf("%w: vector contains a mix of histograms with exponential and custom buckets schemas for metric name", PromQLWarning)
-	IncompatibleCustomBucketsHistogramsWarning = fmt.Errorf("%w: vector contains histograms with incompatible custom buckets for metric name", PromQLWarning)
```
Removed this one.

linasm commented on model/histogram/float_histogram_test.go:
```diff
@@ -2314,6 +2375,9 @@ func TestFloatHistogramAdd(t *testing.T) {
 }
 
 func testHistogramAdd(t *testing.T, a, b, expected *FloatHistogram, expErrMsg string, expCounterResetCollision bool) {
+	require.NoError(t, a.Validate(), "a")
+	require.NoError(t, b.Validate(), "b")
```
It is difficult enough to reason about histogram operations in test cases, and having invalid inputs makes it only harder. Because of that, decided to add validation of test inputs. Had to adjust some existing test cases.

beorn7 commented:
Thanks for this. As said, I do not have capacity for a detailed review right now (but maybe others have). I peeked at it, though, because it is very intriguing.

Just to clarify: From the PromQL perspective, this fixes _all_ the cases where NHCB buckets are mismatched. (While the implementation happens in the Add/Sub methods, it helps for `rate` calculation and everything else, not just for `+` and `-` or the `sum` aggregation.) I'll change the PR's title so that people won't be confused later.

I have a lot of more strategic thoughts about this, which is more around how to release this (before/after declaring NH a stable feature, and how to do it if we do the latter) and if we should have annotation when buckets are reconciled. To not pollute the review of the actual implementation here, I'll add those thoughts to #17255.

linasm commented on model/histogram/float_histogram.go:
```diff
@@ -1346,6 +1363,119 @@ func addBuckets(
+
+	for i < len(boundsA) && j < len(boundsB) {
+		switch {
+		case math.Float64bits(boundsA[i]) == math.Float64bits(boundsB[j]):
```
Sumbitted a PR speculatively: https://github.com/prometheus/prometheus/pull/17284

beorn7 commented:
About reset detection: https://github.com/prometheus/prometheus/blob/9e4d23ddafcdc00021cd8630e78bb819e84ccac9/model/histogram/float_histogram.go#L642-L646 needs an update to do the reconciliation of custom buckets and then detect the resets normally.

linasm commented:
@krajorama I've addressed your comments regarding custom bounds intersection loop. There was another comment about consistency that I'm not sure about (see my response inline).

krajorama commented on model/histogram/float_histogram.go:
```diff
@@ -1331,6 +1366,208 @@ func floatBucketsMatch(b1, b2 []float64) bool {
+	previous *FloatHistogram, currBounds, prevBounds []float64,
+) bool {
+	currIt := h.floatBucketIterator(true, h.ZeroThreshold, h.Schema)
+	prevIt := previous.floatBucketIterator(true, h.ZeroThreshold, h.Schema)
```
I see, however for exponential histograms it's a different story, since in that case we first of all make sure that the current schema is [less or equal](https://github.com/prometheus/prometheus/blob/9a5bccbd4b3b236ac6585324bc0886c3538a1bfe/model/histogram/float_histogram.go#L613) to prev schema and make a check for the zero threshold as well. Then we actually downscale the prev histogram to the current histogram's schema by setting the last parameter in `floatBucketIterator` to the current schema. Long story short, if a high resolution histogram is followed by a lower resolution histogram, we can still do the [counter reset detection](https://prometheus.io/docs/specs/native_histograms/#counter-reset-considerations).

For custom buckets we should not do that in principle since we are doing a different kind of merge of buckets.
So I think for self consistency (not consistency with the exponential case) we should use the zero threshold (always 0 for NHCBs) and schema (always -53 for NHCB) coming from the histogram in question and not from the other.

linasm replied:
Right, hardcoding `0, CustomBucketsSchema` makes most sense for NHCB-specific code. Updated.

beorn7 reviewed (changes requested):
Looks great. Just a minor suggestion to improve the wording of the info annotation.

And maybe there could be tests that calculate non-trivial rates (i.e. one with a bunch of buckets in the result) involving bucket reconciliation, one with a reset and one without a reset?

linasm commented:
@beorn7 I've addressed your feedback:
1. Updated info message as suggested.
2. Added panic in case of non-NHCB arguments, updated the comments, additionally updated method name to make the contract more obvious.
3. Added a few more sophisticated tests. Note that I have used `irate(foo[2m]) * 60` to eliminate any extrapolation effects which were making query results too hard to reason about (at least for me). Hope you won't mind introducing this pattern that seems to be new here (or maybe there is a better approach?).
